### PR TITLE
dev/core#3777 Afform: fix the setting defaults for radio and checkboxes when values are non-string

### PR DESCRIPTION
**Pull request by kurund**
https://lab.civicrm.org/dev/core/-/issues/3777

### CODE REVIEW

**Comment by kurund** [comment]
@colemanw 

> we ought to base it on ctrl.defn.data_type.

There is no such attribute in the field definition hence I have have used the current workaround.

Screenshot from 2023-02-24 18-44-39

Also note that `input_type` is what selected by the user so it does not reflect actual data type.

May be we should add `data_type` to field definition?

**Comment by kurund** [comment]
> May be we should add data_type to field definition?

Problem with this is users can change that too, if it's part of afform. 

I think determining this real-time on client side might be overkill.

Also adding new attributes might work well for new forms but handling upgrade etc might be tricky.

**Comment by colemanw** [comment]
@kurund this would make `data_type` available:

```diff
diff --git a/ext/afform/core/Civi/Afform/FormDataModel.php b/ext/afform/core/Civi/Afform/FormDataModel.php
index 646f2641f5..56a9d06e1a 100644
--- a/ext/afform/core/Civi/Afform/FormDataModel.php
+++ b/ext/afform/core/Civi/Afform/FormDataModel.php
@@ -210,7 +210,7 @@ class FormDataModel {
     $params = [
       'action' => $action,
       'where' => [['name', 'IN', $namesToMatch]],
-      'select' => ['name', 'label', 'input_type', 'input_attrs', 'help_pre', 'help_post', 'options', 'fk_entity', 'required'],
+      'select' => ['name', 'label', 'input_type', 'data_type', 'input_attrs', 'help_pre', 'help_post', 'options', 'fk_entity', 'required'],
       'loadOptions' => ['id', 'label'],
       // If the admin included this field on the form, then it's OK to get metadata about the field regardless of user permissions.
       'checkPermissions' => FALSE,

```

**Comment by kurund** [comment]
@colemanw 

Thanks, I have updated the PR to use data_type.

**Comment by colemanw** [comment]
@kurund this all looks good except, what to do if the value is `null` - your code would convert `null` to `false` or `0` for booleans and integers, but I don't think that's what we want to do.

File: ext/afform/core/ang/af/afField.component.js [hunk 1]
```diff
@@ -142,8 +142,33 @@
         });
       };
 
+      // correct the type for the value, make sure numbers are numbers and not string
+      function correctValueType(value, dataType) {
+        // let's skip type correction for null values
+        if (value == null) {
```
Comment (colemanw) [inline ext/afform/core/ang/af/afField.component.js:148]: Wrong operator for checking `null`, have to use `===` (also our stylechecker wants `NULL` not `null`). How about just do this instead:
```suggestion
        if (is_null(value)) {
```

Comment (colemanw) [inline ext/afform/core/ang/af/afField.component.js:148]: Need to check data type using `===`
```suggestion
        if (value === null) {
```

Comment (kurund) [reply ext/afform/core/ang/af/afField.component.js]: done

**Comment by colemanw** [comment]
Great thanks @kurund

**Comment by xavi-xaloc** [comment]
Already mentioned in https://lab.civicrm.org/dev/core/-/issues/4113#note_149597, but I add here anyway what I found after checking this PR with CiviCRM 5.61.2:

Results now get filtered correctly when a default value is selected and, as for the display of the field:

- Radio button: Works fine with both "yes" and "no" default values.
- Select: Works fine when "yes" is the default value, but doesn't show default value when it is set to "no"
- Checkboxes: Doesn't get checked when default value is set to "yes".

I've filed a separate issue in lab.civicrm, as @kurund suggested: https://lab.civicrm.org/dev/core/-/issues/4465

**Comment by colemanw** [comment]
@xavi-xaloc I've posted https://github.com/civicrm/civicrm-core/pull/27294 to fix the above issues you raised.

**Comment by xavi-xaloc** [comment]
Thank you, @colemanw .I tried to test it, but the online demo of the PR doesn't seem to be accessible

**Comment by colemanw** [comment]
@xavi-xaloc now that it's been merged to master you can test at https://dmaster.demo.civicrm.org/civicrm

**Comment by xavi-xaloc** [comment]
I've tested at dmaster @colemanw , but it doesn't seem to solve of the issues mentioned above. I've created a searchkit for contacts and, in the Formbuilder, I've included three boolean filters:

1. Is deceased: checkbox wtih default value "Yes"
2. Do not phone: radio button with default value "Yes"
3. Do not email: select with default value "No"

But the display only shows default value for radio buttons:
image

Moreover, when the filter is a select, it doesn't applies correctly. In this case, the same results are shown regardless of the value applied for "do not email" filter (it always shows registers with value "no" ):
image

**Comment by colemanw** [comment]
@xavi-xaloc whoops we were both looking at the wrong PR. **This** PR is the original (incomplete) fix and has been merged. But the followup fixes are at https://github.com/civicrm/civicrm-core/pull/27294 and has not been merged so you should use the online demo linked to in that PR.

**Pull request merged**